### ISSUE TITLE
Support "Grounding with Google Search" in Google provider

### ISSUE THREAD

**Issue by wong2** [vercel/ai#3553]
### Feature Description

The Gemini API has a new feature where you can ground results with Google search which reduces hallucinations and provides references to content:
https://ai.google.dev/gemini-api/docs/grounding

**Comment by adolphnov** [comment]
I have tried vertex is valid, it will report an error in Google AI Studio, which does not support this parameter. And note that when used in the vertex, no tools can be included, otherwise an error will also be reported.
I hope this is helpful to you.

**Comment by lgrammel** [comment]
@shaper do we support this with the latest provider versions?

**Comment by shaper** [comment]
The above changes were just published to the package versions in https://github.com/vercel/ai/pull/4081

This enables search grounding support in the Google Generative AI `google` package, similar to what we've had in Google Vertex.

We plan to add support for customizing the dynamic retrieval config soon. At that time I would consider this issue complete. If there is interest in the enterprise data support mentioned above please file that as a separate issue.

**Comment by wottpal** [comment]
Hey there, dope addition! Just played with it a bit and got it to work pretty well. 

Some observations though:

1. It doesn't seem to work with `generateObject`, only `generateText`. At least I was not able to force the model to actually DO search when generating object output. Is that related to mutual exclusiveness with tool calls, you're mentioning above, @shaper? If yes, it should probably be mentioned in the docs.
2. Being able to set the dynamic retrieval threshold is crucial for a lot of use cases IMO. Because in my observations the model sometimes tends to NOT google even though I nudge it towards it in my prompt **a lot**. Would be amazing to see this soon.


Lmk whether I should open separate issues for those if you want to close this one.

**Comment by shaper** [comment]
> Hey there, dope addition! Just played with it a bit and got it to work pretty well.
> 
> Some observations though:
> 
> 1. It doesn't seem to work with `generateObject`, only `generateText`. At least I was not able to force the model to actually DO search when generating object output. Is that related to mutual exclusiveness with tool calls, you're mentioning above, @shaper? If yes, it should probably be mentioned in the docs.

Yes, probably -- I will check this and update docs if needed.

> 2. Being able to set the dynamic retrieval threshold is crucial for a lot of use cases IMO. Because in my observations the model sometimes tends to NOT google even though I nudge it towards it in my prompt **a lot**. Would be amazing to see this soon.

It is on my radar, will try to get to it soon.

> Lmk whether I should open separate issues for those if you want to close this one.

No need, let's keep this one for now.

**Comment by sroze** [comment]
I just created another issue (and a proposed implementation plan) for the "enterprise" feature of grounding in your own data:
https://github.com/vercel/ai/issues/4230

**Comment by YasserDbeis** [comment]
Is there an ETA for being able to enable dynamic retrieval? Dynamic retrieval is a huge potential cost saver.

**Comment by rdvo** [comment]
yeah is there a way to  pass dynamic_threshold as a field?

**Comment by kakaly** [comment]
hey, any ETA on when searchGrounding will work with generateObject?

**Comment by lgrammel** [comment]
@YasserDbeis @rdvo see https://sdk.vercel.ai/providers/ai-sdk-providers/google-generative-ai#dynamic-retrieval for dynamic retrieval settings

@kakaly that should already be possible - what issue did you run into?

**Comment by kakaly** [comment]
@lgrammel - it does not do search grounding with Gemini 2.0 flash. It returns results but those are different from what I see if I see geminis library directly.

**Comment by ezyyeah** [comment]
In aistudio, activating Grounding with Google search, makes structured output disabled. A tooltip shows: `This tool is not compatible with the current active tools.`

The same issue occurs in code. Setting useSearchGrounding to true (even for `gemini-2.5-pro-exp-03-25`) results in `groundingMetadata` being null in the output content, indicating that no search was made. If we set useSearchGrounding to false, `groundingMetadata` contains the details of the search.

Is there a way around this? Like in other places we can have one agent that does the search and writes the output in text, then another makes a structured output of it, but that seems like a waste of tokens.

**Comment by kakaly** [comment]
Yes Google doesn’t support grounding with structured outputs which is why vercel ai sdk doesn’t work either. Right now only work around seems to be that you do 2 inference calls - get the results with grounding inference and then pass the results to another inference to get structured outputs. AI SDk does not support this at the moment afaict.

**Comment by RobSchilderr** [comment]
Is this fixed now [Gemini 3 lets us combine Structured Outputs with built-in tools, including Grounding with Google Search, URL Context, Code Execution, and File Search.](https://ai.google.dev/gemini-api/docs/structured-output?example=recipe#structured_outputs_with_tools)?

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Feature**  
Confidence: **High**

### Rationale
This is primarily a request to add and expand provider capabilities for Google Search grounding, rather than a failure of previously promised behavior.

### Evidence
- The issue explicitly requests adding Google Search grounding support to the Google provider.
- Maintainers added basic grounding support, while identifying dynamic retrieval configuration and grounding metadata as additional unsupported capabilities.
- Later comments request support for combining grounding with structured output after Gemini 3 introduced that capability.
